### ISSUE TITLE
Decouple `Plan` tracing from `BaseWorker`

### ISSUE THREAD

**Issue by karlhigley** [OpenMined/PySyft#3402]
**Is your feature request related to a problem? Please describe.**
Tracing nested `Plans` still depends on stashing state on `BaseWorker` via the `init_plan` attribute. We'd like to decouple `Plan` tracing from the hooks and `BaseWorker`, so that it's self-contained and can trace multiple `Roles` in parallel for `Protocols`.

**Describe the solution you'd like**
 TBD how to do this while keeping nested `Plan` tracing working. Maybe each `Plan` could have an attribute that points to its parent or root-level `Plan`, where the root-level `Plan` points to itself?

**Describe alternatives you've considered**
- Dropping nested `Plan` functionality. Not even sure who uses it for what.

**Additional context**
#3303 removed (most of?) the coupling between `Plan` tracing and the hooks. This change was out of scope for that PR, but is very much related.
